# How to manipulate a pandas dataframe

### 0. Import required modules

In [79]:
import pandas as pd #import pandas library for data manipulation (https://pandas.pydata.org/)
import numpy as np #import the 'NumPy' module for scientific computing with Python (https://numpy.org/)
from datetime import datetime #import the 'datetime' module for manipulating dates and times (https://docs.python.org/3/library/datetime.html)
import random #import random module for pseudo-random number generation

### 1. Read data as pandas dataframe
*see [1. Reading in files using pandas (.csv, .xls, .json)](01_read_files.ipynb)*

In [80]:
df = pd.read_excel('nfl_field_goal_sol.xlsx', sheet_name = 'raw_data').replace(-999, None) #read in nfl_field_goal_sol.xlsx file
df.head() #display the head (first 5 rows) of df

gid     pid fgxp  dist  make  season  postseason  away  \
0  3189  518667   FG    33     1    2012           0     1   
1  3190  518780   FG    35     1    2012           0     0   
2  3190  518806   FG    37     0    2012           0     1   
3  3190  518822   FG    26     1    2012           0     0   
4  3191  518902   FG    43     1    2012           0     0   

                    stadium temp humidity wind-spd           cond    surface  \
0           MetLife Stadium   77       63        5           Fair  FieldTurf   
1             Soldier Field   67       66       13  Partly Cloudy      Grass   
2             Soldier Field   67       66       13  Partly Cloudy      Grass   
3             Soldier Field   67       66       13  Partly Cloudy      Grass   
4  Cleveland Browns Stadium   68       57        9          Sunny      Grass   

  player id          player  
0   DB-0200      Dan Bailey  
1   RG-1500    Robbie Gould  
2   AV-0400  Adam Vinatieri  
3   RG-1500    Robbie Gould  
4   PD-0200     Phil Dawson

In [81]:
df_stadium = pd.read_excel('stadium_df.xlsx', index_col=0) #read in stadium_df.xlsx file 
df_stadium.head() #display the head (first 5 rows) of df_stadium

stadium             stadium-name  stadium-id
0              AT&T Stadium             AT&T Stadium           1
1         Arrowhead Stadium        Arrowhead Stadium           2
2   Bank Of America Stadium  Bank of America Stadium           3
3  Bank of America Stadiium  Bank of America Stadium           3
4   Bank of America Stadium  Bank of America Stadium           3

### 2. Get basic information on dataframe structure

In [82]:
df.info() #display the shape and data types of the 'df' dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4184 entries, 0 to 4183
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   gid         4184 non-null   int64 
 1   pid         4184 non-null   int64 
 2   fgxp        4184 non-null   object
 3   dist        4184 non-null   int64 
 4   make        4184 non-null   int64 
 5   season      4184 non-null   int64 
 6   postseason  4184 non-null   int64 
 7   away        4184 non-null   int64 
 8   stadium     4184 non-null   object
 9   temp        3187 non-null   object
 10  humidity    3019 non-null   object
 11  wind-spd    3103 non-null   object
 12  cond        4159 non-null   object
 13  surface     4184 non-null   object
 14  player id   4184 non-null   object
 15  player      4184 non-null   object
dtypes: int64(7), object(9)
memory usage: 523.1+ KB


#### Result of pandas.info() function... 
The result tells us there are 4184 rows and 16 columns.  
The 'gid', 'pid', 'dist', 'make', 'season', 'postseason', 'away', 'temp' 'humidity' and 'wind-spd' columns are all data-type 'integer'.  
The 'fgxp', 'stadium', 'cond', 'surface', 'player id' and 'player' columns are all data-type 'object'.

### 3. Dealing with *NULL* values

In [83]:
non_null_df = df.dropna() #create new dataframe with all rows containing NULL values removed
non_null_df.info() #display the shape and data types of the 'new_df' dataframe

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2994 entries, 0 to 4181
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   gid         2994 non-null   int64 
 1   pid         2994 non-null   int64 
 2   fgxp        2994 non-null   object
 3   dist        2994 non-null   int64 
 4   make        2994 non-null   int64 
 5   season      2994 non-null   int64 
 6   postseason  2994 non-null   int64 
 7   away        2994 non-null   int64 
 8   stadium     2994 non-null   object
 9   temp        2994 non-null   object
 10  humidity    2994 non-null   object
 11  wind-spd    2994 non-null   object
 12  cond        2994 non-null   object
 13  surface     2994 non-null   object
 14  player id   2994 non-null   object
 15  player      2994 non-null   object
dtypes: int64(7), object(9)
memory usage: 397.6+ KB


#### Result of pandas.info() function...  
This result tells us that after removing rows containing NULL values, the dataframe has reduced in size to 2994 rows.

### 4. Merging/joining dataframes

In [84]:
df2 = pd.merge(df, df_stadium, on='stadium') #create new df (df2) by merging df and df_stadium on the 'stadium' column
df2.head() #display the head (first 5 rows) of df2

gid     pid fgxp  dist  make  season  postseason  away          stadium  \
0  3189  518667   FG    33     1    2012           0     1  MetLife Stadium   
1  3210  522061   FG    45     1    2012           0     1  MetLife Stadium   
2  3210  522138   FG    52     1    2012           0     1  MetLife Stadium   
3  3256  529883   FG    32     1    2012           0     1  MetLife Stadium   
4  3256  529951   FG    41     1    2012           0     1  MetLife Stadium   

  temp humidity wind-spd          cond    surface player id        player  \
0   77       63        5          Fair  FieldTurf   DB-0200    Dan Bailey   
1   73       39        1  Mostly Sunny  FieldTurf   CB-0700  Connor Barth   
2   73       39        1  Mostly Sunny  FieldTurf   CB-0700  Connor Barth   
3   51       84        2          Rain  FieldTurf   PD-0200   Phil Dawson   
4   51       84        2          Rain  FieldTurf   PD-0200   Phil Dawson   

      stadium-name  stadium-id  
0  MetLife Stadium          25  
1  MetLife Stadium          25  
2  MetLife Stadium          25  
3  MetLife Stadium          25  
4  MetLife Stadium          25

### 5. Subsetting & concatenating dataframes

In [85]:
df_2015 = df2.loc[df2['season'] == 2015] #create subset of df for the 2015 season
df_2015.head() #display the head (first 5 rows) of df_2015

gid     pid fgxp  dist  make  season  postseason  away          stadium  \
97   3993  652990   FG    34     1    2015           0     0  MetLife Stadium   
98   4015  656592   FG    38     1    2015           0     0  MetLife Stadium   
99   4015  656605   FG    42     1    2015           0     1  MetLife Stadium   
100  4015  656634   FG    44     1    2015           0     0  MetLife Stadium   
101  4022  657753   FG    35     1    2015           0     0  MetLife Stadium   

    temp humidity wind-spd           cond               surface player id  \
97    77       49        8         Cloudy             FieldTurf   NF-0300   
98    73       47        9  Mostly Cloudy  UBU Speed Series S5M   JB-7100   
99    73       47        9  Mostly Cloudy  UBU Speed Series S5M   MB-4600   
100   73       47        9  Mostly Cloudy  UBU Speed Series S5M   JB-7100   
101   71       60        4          Clear  UBU Speed Series S5M   JB-7100   

          player     stadium-name  stadium-id  
97     Nick Folk  MetLife Stadium          25  
98    Josh Brown  MetLife Stadium          25  
99   Matt Bryant  MetLife Stadium          25  
100   Josh Brown  MetLife Stadium          25  
101   Josh Brown  MetLife Stadium          25

In [86]:
df_2015_ids = df_2015[['gid', 'pid', 'player id', 'stadium-id']] #create subset of df_2015 for ID columns
df_2015_ids.head() #display the head (first 5 rows) of df_2015_ids

gid     pid player id  stadium-id
97   3993  652990   NF-0300          25
98   4015  656592   JB-7100          25
99   4015  656605   MB-4600          25
100  4015  656634   JB-7100          25
101  4022  657753   JB-7100          25

In [87]:
df_concat = pd.concat([df2, df_2015]) #bind df2 and df_2015 together 
df_concat.info() #display info for df_concat

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5120 entries, 0 to 4109
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   gid           5120 non-null   int64 
 1   pid           5120 non-null   int64 
 2   fgxp          5120 non-null   object
 3   dist          5120 non-null   int64 
 4   make          5120 non-null   int64 
 5   season        5120 non-null   int64 
 6   postseason    5120 non-null   int64 
 7   away          5120 non-null   int64 
 8   stadium       5120 non-null   object
 9   temp          3890 non-null   object
 10  humidity      3693 non-null   object
 11  wind-spd      3803 non-null   object
 12  cond          5095 non-null   object
 13  surface       5120 non-null   object
 14  player id     5120 non-null   object
 15  player        5120 non-null   object
 16  stadium-name  5120 non-null   object
 17  stadium-id    5120 non-null   int64 
dtypes: int64(8), object(10)
memory usage: 760.0+ KB


#### Result of pandas.info() function...  
This result tells us that after binding together dataframes 'df2' and 'df_2015', 'df_concat' is now a total of 5120 rows.  
936 rows of df_concat are now duplicates

### 6. Creating & manipulating dataframe columns

In [88]:
df2['dist-meters'] = df2['dist']*0.9144 #calculate new variable 'dist_meters' which converts distance in yards to meters
df2['updated-at'] = datetime.today().strftime('%Y-%m-%d') #create new column to reference when the data was updated
df2.head() #display the head (first 5 rows) of df

gid     pid fgxp  dist  make  season  postseason  away          stadium  \
0  3189  518667   FG    33     1    2012           0     1  MetLife Stadium   
1  3210  522061   FG    45     1    2012           0     1  MetLife Stadium   
2  3210  522138   FG    52     1    2012           0     1  MetLife Stadium   
3  3256  529883   FG    32     1    2012           0     1  MetLife Stadium   
4  3256  529951   FG    41     1    2012           0     1  MetLife Stadium   

  temp humidity wind-spd          cond    surface player id        player  \
0   77       63        5          Fair  FieldTurf   DB-0200    Dan Bailey   
1   73       39        1  Mostly Sunny  FieldTurf   CB-0700  Connor Barth   
2   73       39        1  Mostly Sunny  FieldTurf   CB-0700  Connor Barth   
3   51       84        2          Rain  FieldTurf   PD-0200   Phil Dawson   
4   51       84        2          Rain  FieldTurf   PD-0200   Phil Dawson   

      stadium-name  stadium-id  dist-meters  updated-at  
0  MetLife Stadium          25      30.1752  2022-10-13  
1  MetLife Stadium          25      41.1480  2022-10-13  
2  MetLife Stadium          25      47.5488  2022-10-13  
3  MetLife Stadium          25      29.2608  2022-10-13  
4  MetLife Stadium          25      37.4904  2022-10-13

### 7. Reformatting column data types

In [89]:
df2['updated-at'] = pd.to_datetime(df2['updated-at']) #force column 'Updated_At' to type 'datetime'
df2.info() #display the shape and data types of the 'df' dataframe

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4184 entries, 0 to 4183
Data columns (total 20 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   gid           4184 non-null   int64         
 1   pid           4184 non-null   int64         
 2   fgxp          4184 non-null   object        
 3   dist          4184 non-null   int64         
 4   make          4184 non-null   int64         
 5   season        4184 non-null   int64         
 6   postseason    4184 non-null   int64         
 7   away          4184 non-null   int64         
 8   stadium       4184 non-null   object        
 9   temp          3187 non-null   object        
 10  humidity      3019 non-null   object        
 11  wind-spd      3103 non-null   object        
 12  cond          4159 non-null   object        
 13  surface       4184 non-null   object        
 14  player id     4184 non-null   object        
 15  player        4184 non-null   object  

#### Result of pandas.info() function...
This result shows us that the *updated-at* column has been forced to data-type 'datetime'

### 8. Removing duplicated rows

In section 5 we concatenated/binded dataframes df2 and df_2015 into a singular dataframe. This new dataframe (df_concat) has duplicated rows which prior to aggregation or analysis will need to be removed

In [90]:
df_concat.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5120 entries, 0 to 4109
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   gid           5120 non-null   int64 
 1   pid           5120 non-null   int64 
 2   fgxp          5120 non-null   object
 3   dist          5120 non-null   int64 
 4   make          5120 non-null   int64 
 5   season        5120 non-null   int64 
 6   postseason    5120 non-null   int64 
 7   away          5120 non-null   int64 
 8   stadium       5120 non-null   object
 9   temp          3890 non-null   object
 10  humidity      3693 non-null   object
 11  wind-spd      3803 non-null   object
 12  cond          5095 non-null   object
 13  surface       5120 non-null   object
 14  player id     5120 non-null   object
 15  player        5120 non-null   object
 16  stadium-name  5120 non-null   object
 17  stadium-id    5120 non-null   int64 
dtypes: int64(8), object(10)
memory usage: 760.0+ KB


In [91]:
df_concat.drop_duplicates(inplace = True) #inplace=True replaces the selected dataframe 'df' 
df_concat.info() #display the shape and data types of the 'df' dataframe

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4184 entries, 0 to 4183
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   gid           4184 non-null   int64 
 1   pid           4184 non-null   int64 
 2   fgxp          4184 non-null   object
 3   dist          4184 non-null   int64 
 4   make          4184 non-null   int64 
 5   season        4184 non-null   int64 
 6   postseason    4184 non-null   int64 
 7   away          4184 non-null   int64 
 8   stadium       4184 non-null   object
 9   temp          3187 non-null   object
 10  humidity      3019 non-null   object
 11  wind-spd      3103 non-null   object
 12  cond          4159 non-null   object
 13  surface       4184 non-null   object
 14  player id     4184 non-null   object
 15  player        4184 non-null   object
 16  stadium-name  4184 non-null   object
 17  stadium-id    4184 non-null   int64 
dtypes: int64(8), object(10)
memory usage: 621.1+ KB


#### Result of pandas.info() function...
This result shows us that the pandas.drop_duplicate() function has removed 936 duplicate rows from the dataframe

### 9. Aggregation of data

In [92]:
#--- aggregate total field goals attempted and successful field goals by player
player_fg_success = (
    df.groupby(['player', 'player id'], as_index=False)
    .agg(
        # Get the number of field goal attempts
        total_fg = pd.NamedAgg(column = 'pid', aggfunc = 'count'),
        # Get the actual number of successful kicks
        successful_fg = pd.NamedAgg(column = 'make', aggfunc = 'sum'))
    .sort_values('successful_fg', ascending=False)) #sort aggregation by number of successful field goals

player_fg_success.head() #display aggregation

player player id  total_fg  successful_fg
30  Stephen Gostkowski   SG-0800       198            179
17       Justin Tucker   JT-3950       195            176
31     Steven Hauschka   SH-0400       187            169
0       Adam Vinatieri   AV-0400       176            154
19        Mason Crosby   MC-3000       180            149

In [93]:
#--- aggregate total field goals attempted and successful field goal rate by player
player_fg_success_rate = (
    df.groupby(['player', 'player id'], as_index=False)
    .agg(
        # Get the number of field goal attempts
        total_fg = pd.NamedAgg(column = 'pid', aggfunc = 'count'),
        # Get the actual number of successful kicks
        successful_fg_rate = pd.NamedAgg(column = 'make', aggfunc = 'mean'))
 .sort_values('successful_fg_rate', ascending=False)) #sort aggregation by % of successful field goals

player_fg_success_rate.head() #display aggregation

player player id  total_fg  successful_fg_rate
15          Josh Brown   JB-7100       110            0.918182
30  Stephen Gostkowski   SG-0800       198            0.904040
31     Steven Hauschka   SH-0400       187            0.903743
6        Chris Boswell   CB-2275        72            0.902778
17       Justin Tucker   JT-3950       195            0.902564